ffi/rtc.lua

```diff
@@ -0,0 +1,196 @@
+--[[--
+Module for interfacing with the RTC (real time clock).
+
+This module provides the ability to schedule wakeups through RTC.
+
+@module ffi.rtc
+]]
+
+local ffi = require("ffi")
+local bor = bit.bor
+local C = ffi.C
+
+local dummy = require("ffi/posix_h")
+local dummy = require("ffi/rtc_h")
+
+-----------------------------------------------------------------
+
+local RTC = {
+    _wakeup_scheduled = false,
+    _wakeup_scheduled_ptm = nil,
+}
+
+--[[
+Adds seconds to epoch.
+
+@int seconds_from_now
+@treturn cdata epoch
+--]]
+function RTC:secondsFromNowToEpoch(seconds_from_now)
+    local t = ffi.new("time_t[1]")
+    t[0] = C.time(nil)
+    t[0] = t[0] + seconds_from_now
+    local epoch = C.mktime(C.localtime(t))
+    return epoch
+end
+
+--[[
+Set wakeup alarm.
+
+@int epoch If you want to set to an amount of time from now, process your value with @{secondsFromNowToEpoch}
+@enabled bool Whether the call enables or disables the alarm. Defaults to true.
+
+@treturn bool Success.
+@treturn re Error code (if any).
+@treturn err Error string (if any).
+--]]
+function RTC:setWakeupAlarm(epoch, enabled)
+    enabled = (enabled ~= nil) and enabled or true
+
+    local ptm = C.gmtime(ffi.new("int[1]", epoch))
+    self._wakeup_scheduled_ptm = ptm
+
+    local wake = ffi.new("struct rtc_wkalrm")
+    wake.time.tm_sec = ptm.tm_sec
+    wake.time.tm_min = ptm.tm_min
+    wake.time.tm_hour = ptm.tm_hour
+    wake.time.tm_mday = ptm.tm_mday
+    wake.time.tm_mon = ptm.tm_mon
+    wake.time.tm_year = ptm.tm_year
+    -- wday, yday, and isdst fields are unused by Linux
+    wake.time.tm_wday = -1
+    wake.time.tm_yday = -1
+    wake.time.tm_isdst = -1
+
+    wake.enabled = enabled and 1 or 0
```

Comment:
If you want to disable it, you can provide any time value, including 0? (= it doesn't matter?)

Reply:
Yeah, setting it to a time in the past would effectively disable it, and I don't think I actually tried what happens when you combine disabled with a time in the future. It can probably be removed.

Also see http://man7.org/linux/man-pages/man4/rtc.4.html